iceoryx_utils/include/iceoryx_utils/internal/cxx/function_ref.inl

```diff
@@ -60,12 +60,8 @@ function_ref<ReturnType(ArgTypes...)>::operator=(function_ref<ReturnType(ArgType
 template <class ReturnType, class... ArgTypes>
 inline ReturnType function_ref<ReturnType(ArgTypes...)>::operator()(ArgTypes... args) const noexcept
 {
-    if (!m_pointerToCallable)
-    {
-        // Callable was called without user having assigned one beforehand
-        std::cerr << "Empty function_ref invoked" << std::endl;
-        std::terminate();
-    }
+    // Expect that a callable was assigned beforehand
+    cxx::Expects((m_pointerToCallable != nullptr) && "Empty function_ref invoked");
```

Comment:
In `method_callback.hpp` we adjusted the signature to return the `ReturnType` packed inside an expected like:
```cpp
template <typename... MethodArguments>
expected<ReturnValue, MethodCallbackError> operator()(MethodArguments&&... args)
```

This error handling is much cleaner since the user is informed and this error is actually something where we can recover from easily. So let the user decide what to do.

If we go the way with the `Expects` the user has to be aware of this and has to read the documentation carefully - which most of the time does not happen. But if we return an expected, the user knows that something can go south, since the compiler enforces an error handling here.
This can reduce unexpected failures in user code.

Reply:
I definitely see your point. We should try to use `expected` wherever possible. However, IMHO it's more important to be close to the `function_ref` proposal of the standard committee. I would try to avoid adding more dependencies like the `expected` here but rather try to be aligned with the STL. This has the usability downsides that you mentioned, but for the sake of compatibility we should strive for plain and simple building blocks in utils/hoofs.

Added a hint in doxygen.